thirtyfour/src/extensions/query/conditions.rs

```diff
@@ -23,483 +20,308 @@ pub(crate) fn negate(result: WebDriverResult<bool>, ignore_errors: bool) -> WebD
+                ignore_errors: bool,
+            ) -> impl ElementPredicate
+            where
+                N: Needle + Clone + Send + Sync + 'static,
+            {
+                move |elem: &WebElement| {
+                    let elem = elem.clone();
+                    let name = [<$single _name>].clone();
+                    let value = value.clone();
+                    elem_matches!(@[<inner_single_ $name>] elem, name, ignore_errors, $field, |x| value.is_match(&x))
                 }
             }
-            Ok(true)
-        })
-    })
-}
 
-/// Predicate that returns true for elements that do not have any of the specified CSS properties
-/// with the specified values.
-/// See the `Needle` documentation for more details on text matching rules.
-pub fn element_lacks_css_properties<S, N>(
-    desired_css_properties: &[(S, N)],
-    ignore_errors: bool,
-) -> ElementPredicate
-where
-    S: Into<String> + Clone,
-    N: Needle + Clone + Send + Sync + 'static,
-{
-    let desired_css_properties: Vec<(String, N)> =
-        desired_css_properties.iter().cloned().map(|(a, b)| (a.into(), b)).collect();
-    Box::new(move |elem| {
-        let desired_css_properties = desired_css_properties.clone();
-        Box::pin(async move {
-            for (css_property_name, value) in &desired_css_properties {
-                match elem.css_value(css_property_name).await {
-                    Ok(x) => {
-                        if value.is_match(&x) {
-                            return Ok(false);
-                        }
-                    }
-                    Err(e) => return handle_errors(Err(e), ignore_errors),
+            #[doc = concat!("Predicate that returns true for elements that do not contain the specified ",stringify!($plural)," with the specified")]
+            /// value. See the `Needle` documentation for more details on text matching rules.
+            pub fn [<element_lacks_ $single>]<N>(
+                [<$single _name>]: Arc<str>,
+                value: N,
+                ignore_errors: bool,
+            ) -> impl ElementPredicate
+            where
+                N: Needle + Clone + Send + Sync + 'static,
+            {
+                move |elem: &WebElement| {
+                    let elem = elem.clone();
+                    let name = [<$single _name>].clone();
+                    let value = value.clone();
+                    elem_matches!(@[<inner_single_ $name>] elem, name, ignore_errors, $field, |x| !value.is_match(&x))
                 }
             }
-            Ok(true)
-        })
-    })
+        }
+    };
+
+    ([$name: ident] $field: ident => $single: ident, $plural: ident) => {
+        elem_matches!(many   [$name] $field => $plural);
+        elem_matches!(single [$name] $field => $single);
+    }
 }
+
+elem_matches!([opt] attr => attribute, attributes);
+elem_matches!([opt] prop => property,  properties);
+elem_matches!([css] css_value => css_property,  css_properties);
```

Comment:
I'm not sure the complexity cost of the macro is worth it to save some copy/paste, especially since these lines rarely change. This also adds to the compile time. I'll leave this in for now.

Reply:
I mean I don't think the compile time is that much of an add, but I just saw that it was  basically duplicated for those three, the rest like text and value I left in, but those confused me when I was removing the boxing for the element predicate change